### PR TITLE
Improving MDN Javascript Fathead

### PR DESCRIPTION
**Pull request by bfmags**
Fixing #280 [WIP]

- [ ] MDN JavaScript: fetch the errors reference articles and parse them
https://github.com/duckduckgo/zeroclickinfo-fathead/issues/280#issuecomment-241023585

----------------------------------
IA Page: https://duck.co/ia/view/mdnjs
Forum topic: [Improve MDN JavaScript Fathead](https://forum.duckduckhack.com/t/improve-mdn-javascript-fathead/155)

This pull request is part of the [Programming Mission](https://forum.duckduckhack.com/t/duckduckhack-programming-mission-overview/53): help us improve the results for [JavaScript related searches](https://forum.duckduckhack.com/t/javascript-search-overview/94)!

### CODE REVIEW

**Comment by daxtheduck** [comment]
Metadata placeholder post
            
*If you haven't already, please edit your first comment to include a link to the IA page at the bottom.*
            
---
            
**How to find your IA page**
- New Instant Answer
  -- follow the steps [here](https://duck.co/ia/new_ia) to create your IA page.
              
- Existing Instant Answer
  -- find your IA page from this [list](https://duck.co/ia)

**Comment by MariagraziaAlastra** [comment]
@bfmags I just deployed your changes to beta!
See an example query here: [js bad_octal](https://beta.duckduckgo.com/?q=js+bad_octal&t=h_&ia=about)

So far it looks good. As discussed on slack, when you'll tackle the redirects you might want to remove plurals and underscores, and as you said you can use regexes for this purpose.

Also, minor but I think it's better to fix it: when I parse I get the error:
<pre>SyntaxError: Using //@ to indicate sourceURL pragmas is deprecated. Use //# instead</pre>

Thanks a lot for working on this!

**Comment by MariagraziaAlastra** [comment]
Actually @bfmags don't mind that error: seems to be gone now. I think we can ignore, unless it appears again.

Thanks!

**Comment by MariagraziaAlastra** [comment]
@bfmags It's on beta again! No need for underscores now to trigger the redirects, nice :smile:

- [js bad octal](https://beta.duckduckgo.com/?q=js+bad++octal&t=h_&ia=about)
- [js history api](https://beta.duckduckgo.com/?q=js+history+api&t=h_&ia=about) (fixed)

As we said on Slack, it might be good to make redirects using synonims. Eg for "bad octal" some example redirects would be "invalid octal", "octal not valid", "not a legal octal", etc.

**Comment by bfmags** [comment]
some progress at making redirects using synonyms - need some help with the list of words:synonyms 

e.g. using ["bad", "not legal", "invalid", "not a valid"]
```
 errors bad octal	R	SyntaxError: "x" is not a legal ECMA-262 octal constant										
+errors not legal octal	R	SyntaxError: "x" is not a legal ECMA-262 octal constant										
+errors invalid octal	R	SyntaxError: "x" is not a legal ECMA-262 octal constant										
+errors not a valid octal	R	SyntaxError: "x" is not a legal ECMA-262 octal constant										
 bad octal	R	SyntaxError: "x" is not a legal ECMA-262 octal constant										
+not legal octal	R	SyntaxError: "x" is not a legal ECMA-262 octal constant										
+invalid octal	R	SyntaxError: "x" is not a legal ECMA-262 octal constant										
+not a valid octal	R	SyntaxError: "x" is not a legal ECMA-262 octal constant	
```

latest output.txt diff 
https://gist.github.com/bfmags/2a0cce8a00a6388d74a3a821c3515cd7

File: lib/fathead/mdnjs/parse.py [hunk 1]
```diff
@@ -267,6 +270,8 @@ def __init__(self, writer):
         self.WEBAPI_CLASS_WORDS = ['Window', 'Navigator', 'MouseEvent',
                             'KeyboardEvent', 'GlobalEventHandlers', 'Element',
                             'Node', 'Event', 'Selection']
+        # for Error pages
+        self.ERROR_SYNONYMS = [ ["bad", "not legal", "invalid", "not a valid"] ]
```
Comment (MariagraziaAlastra) [inline lib/fathead/mdnjs/parse.py]: 👍

**Comment by MariagraziaAlastra** [comment]
@bfmags it's on beta!
- [js invalid octal](https://beta.duckduckgo.com/?q=js+invalid+octal&t=h_&ia=about)
- [js not a valid octal](https://beta.duckduckgo.com/?q=js+not+a+valid+octal&t=h_&ia=software)
- [js bad array length](https://beta.duckduckgo.com/?q=js++bad+array+length&t=h_&ia=about)

Looks mostly good, although I don't see the main article for "SyntaxError: JSON.parse: bad parsing", so all the redirects [here](https://gist.github.com/bfmags/2a0cce8a00a6388d74a3a821c3515cd7#file-output-diff-synonyms-L1042-L1048) don't trigger on beta.

Also, what about adding "not valid" and "illegal" among the synonims?

Thanks!

**Comment by MariagraziaAlastra** [comment]
As agreed on Slack, I'm going to merge and release these changes and any further improvements can be done on a new PR.

Thanks @bfmags! :+1:

**Pull request merged**